consensus/istanbul/types.go

```diff
@@ -343,6 +425,65 @@ type Message struct {
 	Msg       []byte
 	Address   common.Address // The sender address
 	Signature []byte         // Signature of the Message using the private key associated with the "Address" field
+
+	// The below fields are not serializable since they are private, they are
+	// set when calling Message.FromPayload, only one will be set in any
+	// instance, which is set depends on the Message.Code.
+	committedSubject    *CommittedSubject
+	prePrepare          *Preprepare
+	prepare             *Subject
+	roundChange         *RoundChange
+	queryEnode          *QueryEnodeData
+	forwardMessage      *ForwardMessage
+	enodeCertificate    *EnodeCertificate
+	versionCertificates []*VersionCertificate
+	valEnodeShareData   *ValEnodesShareData
+}
+
+// NewMessage constructs a message with the innerMessage instance and the
+// sender address, it will panic if innerMessage is not a recognised inner
+// message type.
+func NewMessage(innerMessage interface{}, sender common.Address) *Message {
+	message := &Message{
+		Address: sender,
+	}
+	switch t := innerMessage.(type) {
+	case *Preprepare:
+		message.Code = MsgPreprepare
+		message.prePrepare = t
+	case *Subject:
+		message.Code = MsgPrepare
+		message.prepare = t
+	case *CommittedSubject:
+		message.Code = MsgCommit
+		message.committedSubject = t
+	case *RoundChange:
+		message.Code = MsgRoundChange
+		message.roundChange = t
+	case *QueryEnodeData:
+		message.Code = QueryEnodeMsg
+		message.queryEnode = t
+	case *ForwardMessage:
+		message.Code = FwdMsg
+		message.forwardMessage = t
+	case *EnodeCertificate:
+		message.Code = EnodeCertificateMsg
+		message.enodeCertificate = t
+	case []*VersionCertificate:
+		message.Code = VersionCertificatesMsg
+		message.versionCertificates = t
+	case *ValEnodesShareData:
+		message.Code = ValEnodesShareMsg
+		message.valEnodeShareData = t
+	default:
+		panic(fmt.Sprintf("attempt to construct Message unrecognised inner message type %T", t))
+	}
+	bytes, err := rlp.EncodeToBytes(innerMessage)
+	if err != nil {
+		panic(fmt.Sprintf("attempt to serialise inner messgae of type %T failed", innerMessage))
```

Comment:
Are we absolutely sure this cannot occur under any normal circumstances? If there is any way for a crafted message from an attacker to trigger this code path, then this becomes a DoS attack. Looking through the `rlp` codebase, there seem to be a number of ways this can result in an error. Most of them involve unserializable types, and would be caught on the first invocation of this function in a test or testnet, but it's unclear if that is always the case, or if a crafted message could surface an error that went otherwise unnoticed.

Another option here is obviously to return the `error` and allow the caller to handle it. I can see the appeal of cleanliness from this approach, but it does make me a bit nervous. I am curious to hear more about why we should panic here.

Reply:
I'm rarely absolutely sure of anything, but I'd say in this case I'm pretty sure.

At a high level the intuition is that we should never be passing random types to NewMessage, since nodes only use  NewMessage when they need to convey something to the network and at that point they would call NewMessage with a specific type.

If you look at where NewMessage is being called you can see that it is always called with specific types, (i.e there is never a case where we pass a parameter of type `interface{}` to NewMessage, which means that regardless of what an attacker may be able to do, we can be sure that the set of types that NewMessage can be called with cannot be modified.

How can we be sure that those types are supported? We can just look at the code and make sure that they are supported, or we can have tests that that exercise all usages of NewMessage, this is preferable.

The normal case in  which this would be triggered is where NewMessage is called with an unsupported type, which implies that there is an error in our code (i.e there is no way to recover from this error without changing the code) in those cases I think panic is a reasonable course of action.

Code can panic, its an unavoidable part of the go language, there is no way to totally avoid them in your code. I think we should just ensure that we trust our tests to uncover any potential panics. It seems that this is the approach we have used thus far successfully, given that any pointer access, array/slice access and channel closing/sending can result in a panic. So I think we can use that same approach in this case.

Reply:
I don't fully agree with the general principle that the best answer to a code error is always to `panic`, but I agree in this case. Thanks for the explanation.